### PR TITLE
feat: Postgres

### PR DESCRIPTION
**Pull request by HarryET**
# Description
Adds Postgres as the primary datastore

## How Has This Been Tested?
Locally tested

## Due Diligence

* [x] Breaking change
* [x] Requires a documentation update
* [x] Requires a e2e/integration test update

### CODE REVIEW

**Review by arein — DISMISSED** [review]
Readme instructions on how to run the db locally would be great or even better a docker compose 

Maybe a follow up PR

File: src/store.rs [hunk 1]
```diff
@@ -1,35 +1,52 @@
...
+        self.execute(query).await?;
+
         Ok(())
     }
 
-    fn get_client(&self, id: &str) -> error::Result<Option<&Client>> {
-        let client = self.get(&K::from(id.to_string()));
-        Ok(client)
+    async fn get_client(&self, _id: &str) -> error::Result<Option<Client>> {
+        // TODO Get Clients
```
Comment (heilhead) [inline src/store.rs]: Shouldn't this be implemented before merging?

**Review by heilhead — APPROVED** [review]
Looks good except the unimplemented part of the SQL storage. Should we address that before merging, what's the plan?

**Review by heilhead — DISMISSED** [review]

File: docker-compose.yml [hunk 2]
```diff
@@ -3,5 +3,20 @@ services:
...
+  echo-server:
+      build: .
+      ports:
+        - "3000:3000"
+      environment:
+        - PORT=3000
+        - LOG_LEVEL=INFO
+        - DATABASE_URL=postgres://postgres:root@postgres:5432/postgres
+        - TELEMETRY_ENABLED=true
+        - TELEMETRY_GRPC_URL=http://jaeger:4317
```
Comment (arein) [inline docker-compose.yml:22]: neat

File: src/main.rs [hunk 3]
```diff
@@ -46,7 +36,20 @@ async fn main() -> error::Result<()> {
...
+        .expect("database url is missing");
+
+    let store = PgPoolOptions::new()
+        .max_connections(5)
+        .connect(database_url)
+        .await?;
+
+    // Run database migrations. `./migrations` is the path to migrations, relative to the root dir (the directory
+    // containing `Cargo.toml`).
+    sqlx::migrate!("./migrations").run(&store).await?;
```
Comment (arein) [inline src/main.rs:51]: what happens if we run multiple instances of the app?

File: src/providers/noop.rs [hunk 4]
```diff
@@ -41,12 +39,12 @@ impl NoopProvider {
...
-#[cfg(test)]
-impl NoopProvider {
-    /// Get notifications for a specific token
-    pub fn get_notifications(&mut self, token: String) -> &Vec<String> {
-        self.bootstrap(token.clone());
-
-        self.notifications.get(&token).unwrap()
-    }
-}
+// #[cfg(test)]
```
Comment (arein) [inline src/providers/noop.rs:42]: obvio no dead code - cool to fix later

**Review by arein — APPROVED** [review]

File: src/main.rs (hunk 3 above)
Comment (HarryET) [reply src/main.rs:51]: It has a migrations table, it will ignore as long as they are in-sync

File: src/providers/noop.rs (hunk 4 above)
Comment (HarryET) [reply src/providers/noop.rs:42]: Yeah, I'll add it too the refactor issue

File: docker-compose.yml (hunk 2 above)
Comment (HarryET) [reply docker-compose.yml:22]: :)

File: src/main.rs (hunk 3 above)
Comment (heilhead) [reply src/main.rs:51]: > what happens if we run multiple instances of the app?

There's a [locking mechanism](https://github.com/launchbadge/sqlx/blob/main/sqlx-core/src/migrate/migrator.rs#L70-L81) (enabled by default), which should prevent race conditions.

**Pull request merged**